### ISSUE TITLE
Questions about TensorCache implementation details

### ISSUE THREAD

**Issue by aleSuglia** [allenai/allennlp#5030]
Hi there,

I am currently using your `TensorCache` implementation for storing visual features. However, there are a few implementation details that are not clear to me. I'll write a list here:

- Why is the number of `max_readers=cpu_count * 2` and  `max_spare_txns=cpu_count * 2`? I believe this will incredibly limit the concurrent access to the database.
- Why is the additional cache `self.cache_cache` required? It's not clear to me from the comment you have. 
- I feel like when saving a `Tensor`,  after detaching, we should be moving it to CPU using `.cpu()`. This should solve the issue that I raised a few days ago I believe (and fixed by @dirkgr).

I personally keep having issues with your `TensorCache` in a multiprocessing scenario. I can see that in your allennlp-models package you never use multiple workers so maybe that's why you've never had issues. Another error that I frequently have is the following: `MDB_BAD_RSLOT: Invalid reuse of reader locktable slot` which seems to be related to the locking mechanism. Let me know what are your thoughts on this. Thanks for your help!

**Comment by dirkgr** [comment]
* `max_readers` and `max_spare_txns` are just heuristically chosen. In our scenario you'd almost never have more readers than CPUs. If your scenario is different, there is no problem with setting those values higher.
* `self.cache_cache` is required for the situation when you get the same tensor out of the `TensorCache` twice. We want to make sure you get the same tensor object each time. Basically, we want to make sure that `tensor_cache["foo"] == tensor_cache["foo"]`, and even `id(tensor_cache["foo"]) == id(tensor_cache["foo"])`. If we didn't have `cache_cache`, every lookup into `TensorCache` would create a new copy of the tensor.
* In my version of torch, `map_location="cpu"` already makes sure that the resulting tensor lives on the CPU. I am not sure what's up in your version that this does not work. I've just merged #5026, which makes sure the tensors are CPU tensors already when they are written. Do you think we need more?

We do actually have multiple workers in many scenarios. This is one of the ways you can configure `MultiprocessDataLoader`. `TensorCache` was written specifically for this scenario. If it doesn't work for you, we have to fix it!

`MDB_BAD_RSLOT` usually means the code is nesting multiple read transactions. I can't think of a scenario where this would happen, but maybe I'm missing something. Are you using Python threads? Can you reproduce this in a small example? I'd be quite interested in fixing it.

**Comment by aleSuglia** [comment]
Thanks @dirkgr for your answers. I only have one question about the second point. Even if you create a copy of the object at that time, wouldn't that be garbage collected the moment the `Instance` object goes of the scope? Maybe I'm missing some details about the LMDB design.

Regarding the multiprocess scenario. Let me give you more context about it. I have N different dataset readers that I combine using an `InterleavingDatasetReader`. Each dataset reader is a subclass of a `VisionReader` dataset reader. Each dataset reader implements the `_read()` and calls `self.shard_iterable()` on the dataset that I'm reading from. Have you tried this scenario?

**Comment by dirkgr** [comment]
> Even if you create a copy of the object at that time, wouldn't that be garbage collected the moment the Instance object goes out of the scope? Maybe I'm missing some details about the LMDB design.

Maybe, maybe not. We want it to be safe to say `tensor_cache["key"]` multiple times even in the same function. Let's say you're using it in a loop. It might make 1000 copies of that tensor very quickly. There is no guarantee when any of these get collected. This has nothing to do with LMDB. It would be an issue with any object serialization.

As for the maximum number of readers, I think we should change the number of reader slots it creates from `2 * #cpus` to `max(127, 2 * #cpus)`. Would that be enough for your case?

I think we should leave `max_spare_txns` at its default of `1`, i.e., not specify it at all. It should have always been `1`. That setting doesn't do what I thought it does.

Would you be willing to put your fixed to `InterleavingDataset` into a PR, together with the other changes? I think they are quite useful, and the community would benefit from them. I'm happy to walk you through those steps.

I hope that these changes together will fix the `MDB_BAD_RSLOT` issue, but if not, we'll tackle that next.

**Comment by aleSuglia** [comment]
When the second epoch is completed the code hangs and this is the stacktrace:
```
0 | 2021-03-02 00:46:12,994 - WARNING - allennlp.training.trainer - Worker 0 completed its entire epoch (training).                      │
^CTraceback (most recent call last):                                                                                                     │
  File "/scratch/anaconda3/envs/devel/bin/allennlp", line 8, in <module>                                                                 │
    sys.exit(run())                                                                                                                      │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/__main__.py", line 34, in run                                 │
    main(prog="allennlp")                                                                                                                │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/__init__.py", line 119, in main                      │
    args.func(args)                                                                                                                      │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/train.py", line 110, in train_model_from_args        │
    train_model_from_file(                                                                                                               │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/train.py", line 170, in train_model_from_file        │
    return train_model(                                                                                                                  │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/train.py", line 308, in train_model                  │
    mp.spawn(                                                                                                                            │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 199, in spawn                    │
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')                                                         │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 157, in start_processes          │
    while not context.join():                                                                                                            │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 75, in join                      │
    ready = multiprocessing.connection.wait(                                                                                             │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/connection.py", line 931, in wait                                    │
    ready = selector.select(timeout)                                                                                                     │
  File "/scratch/anaconda3/envs/devel/lib/python3.8/selectors.py", line 415, in select                                                   │
    fd_event_list = self._selector.poll(timeout)                                                                                         │
KeyboardInterrupt
UserWarning: resource_tracker: There appear to be 10 leaked semaphore objects to clean up at shutdown
```
This seems to happen right before starting the validation step for the second epoch. And inspecting the workers logs I've found this:

```
WARNING - allennlp.training.trainer - Worker 1 finishing validation early! This implies that there is an imbalance in your validation data across the workers and that some amount of it will be ignored. A small amount of this is fine, but a major imbalance should be avoided. Note: This warning will appear unless your data is perfectly balanced.
```
First of all, why is this happening? Is it due to the laziness of the data loader? Second, how can I disable this behaviour during validation? I want all the instances to be processed!

**Comment by epwalsh** [comment]
> First of all, why is this happening? Is it due to the laziness of the data loader? Second, how can I disable this behaviour during validation? I want all the instances to be processed!

The trainer requires there to be an equal number of batches (both for training and validation) across distributed workers. So, for example, if your batch size is 1, you are training on 4 GPUs, and you only have 7 instances, then 3 instances will have to be ignored.

> setting `batches_per_epoch` in a non-multi-processing setup at least prevents the program to hang. I feel like there is some edge-case in the way an entire epoch is processed.

I would recommend against setting `batches_per_epoch` in your validation loader. This could result in having a different validation set each epoch.

> In my version, seems that having multiple workers breaks the feature extraction process because every dataset reader tries to instantiate the FasterRCNN model in a nested sub-process and I think this breaks PyTorch multiprocessing (not sure 100% about this statement).

This is true (every worker is instantiating the region detector), and this requires CUDA to be initialized. As the error says, you need to use the "spawn" start method in order to initialize CUDA in workers. So just set `data_loader.start_method` to `"spawn"`.

**Comment by dirkgr** [comment]
So it seems like @aleSuglia's validation set produces one batch too many/too few in one of the workers, and thus they go out of sync?

**Comment by epwalsh** [comment]
> So it seems like @aleSuglia's validation set produces one batch too many/too few in one of the workers, and thus they go out of sync?

The trainer prevents itself from getting out of sync by stopping early as soon as one worker has completely its last batch. Hence the warning. So that doesn't explain why the code hangs.

The hang could be caused by a bug in whatever metric you're using if the metric can't handle unequal batch sizes across distributed workers. I recently found (and fixed) a bug like that in the SQuAD EM+F1 metric which was causing a hang in the same scenario as you describe. (https://github.com/allenai/allennlp-models/commit/9472904acd94f6fca168713a6183ed6ecd2449cf)

**Comment by dirkgr** [comment]
@aleSuglia, what are the metrics you're using at validation time? If we have a bug in those, that would be quite important to fix.

**Comment by aleSuglia** [comment]
I'm using BLEU and ROUGE. Let me know if you need more information about my setup. I would love to get this sorted!

**Comment by epwalsh** [comment]
I just looked through the implementation of BLEU and ROUGE carefully. I don't think there's any issue there.

@aleSuglia how are you calling those metrics? I'm assuming you only call them once per batch with the batch tensors (this is how they're supposed to be called). But if instead you called the metrics `batch_size` times per batch, that would certainly cause a dead lock. If my question isn't clear, could you just post the snippet of your code that calls those metrics?

**Comment by aleSuglia** [comment]
Oh no I call the metrics once per batch and only during validation (mainly because I have a generative model).

**Comment by epwalsh** [comment]
Hmm okay. Can you try to widdle your code down to a minimal reproducible example? Or if you can find the exact line where the hang occurs, that should give me enough information to figure out what's going on.

**Comment by aleSuglia** [comment]
So my total number of instances is 2276328 (that I computed using the `count-instances` utility). I have 5 dataset readers that are used in my training using `InterleavingDatasetReader` (with round robin scheme). Three of these datasets have a size which is around 100k. The other two dataset reader generate around 900k. As you can see, the dataset is heavily imbalanced in terms of number of instances. I'm currently computing the dataset sizes for each of them so I can give you a better figure. 

All my subclasses of `VisionReader` use the following `_read()` method:
```
    def _read(self, file_path) -> Iterable[Instance]:
        # we read the JSONL data first
        with gzip.open(file_path, "rb") as in_file:
            j_reader = jsonlines.Reader(in_file)
            caption_dicts = list(
                self.shard_iterable(
                    caption_dict for caption_dict in j_reader
                )
            )

        processed_images: Iterable[Optional[Tuple[Tensor, Tensor]]]
        if self.produce_featurized_images:
            # It would be much easier to just process one image at a time, but it's faster to process
            # them in batches. So this code gathers up instances until it has enough to fill up a batch
            # that needs processing, and then processes them all.
            filenames = [f"{question_dict['image_filename']}.jpg" for question_dict in caption_dicts]
            try:
                processed_images = self._process_image_paths(
                    self.images[filename] for filename in filenames
                )
            except KeyError as e:
                missing_filename = e.args[0]

                raise KeyError(
                    missing_filename,
                    f"We could not find an image with the name {missing_filename}. "
                    "Because of the size of the image datasets, we don't download them automatically. "
                    "Please download the images from"
                    "https://cocodataset.org/, "
                    "extract them into a directory, and set the image_dir parameter to point to that "
                    "directory. This dataset reader does not care about the exact directory structure. It "
                    "finds the images wherever they are.",
                )
        else:
            processed_images = [None] * len(caption_dicts)

        for caption_dict, processed_image in zip(caption_dicts, processed_images):
            # we will delegate the instance generation process to the subclass
            instance = self.text_to_instance(caption_dict, processed_image)
            if instance is not None:
                yield instance
```  
Apart from my actual data, I am using specifically AllenNLP components as they are so I don't think the issue is in my code. I believe is more like an issue in terms of the way multiprocessing is dealing with the huge amount of instances.

**Comment by dirkgr** [comment]
I don't know where we're at with this now.
> As for the maximum number of readers, I think we should change the number of reader slots it creates from 2 * #cpus to max(127, 2 * #cpus). Would that be enough for your case?

> I think we should leave max_spare_txns at its default of 1, i.e., not specify it at all. It should have always been 1. That setting doesn't do what I thought it does.

Did you try these fixes? Did they fix something?

I don't think there is anything wrong with your reader.

`InterleavingDatasetReader` will instantiate one copy of `FasterRCNN` for each of the child readers. Are you running out of GPU memory because of this?

What happens when you don't set `num_workers`? You should get far fewer processes trying to do things at the same time if you do that. Since you're pre-processing images, and doing that is already batched inside each reader, there is not much benefit to also use so many workers.

**Comment by aleSuglia** [comment]
@dirkgr I have tried to apply those changes but I still have this error: `MDB_BAD_RSLOT: Invalid reuse of reader locktable slot`. Any ideas about how to fix it?

Yes, the `InterleavingDatasetReader` has a small batch size (equal to 2) at the moment so that I do not run out of memory. I thought it would be beneficial to have multiple workers so that, once the features have been properly generated, the training would be faster. I still have issues with the AllenNLP data loader which I believe has some weird bug related to the lazy dataset implementation. Any ideas about it?

**Comment by dirkgr** [comment]
Is it possible that you are creating a `TensorCache` before `fork`ing the process?

Have you tried @epwalsh's suggestion to "set `data_loader.start_method` to 'spawn'"?

`num_workers` has very little to do with the batch size. You should not set `num_workers` until you see that data loading is a bottleneck. Pre-processing images is already parallelized without setting `num_workers`. There is no benefit to having more than one worker per GPU. When your cache is full, you will likely see that data loading is still not a bottle neck, because now all it does is read tensors from disk, which does not take long.

**Comment by aleSuglia** [comment]
Sorry for the delay in my answer.

As I'm creating a dataset reader that extends `VisionReader`, I don't think I'm creating a `TensorCache` before `fork`ing the process. When extending the `VisionReader`, I'm only changing the `_read` and `text_to_instance` method. 

Regarding @epwalsh suggestion of using "spawn". If I tried that with 4 workers I get the following stacktrace:

```
-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/torch/multiprocessing/spawn.py", line 19, in _wrap
    fn(i, *args)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/train.py", line 466, in _train_worker
    metrics = train_loop.run()
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/commands/train.py", line 528, in run
    return self.trainer.train()
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/training/trainer.py", line 930, in train
    metrics, epoch = self._try_train()
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/training/trainer.py", line 962, in _try_train
    train_metrics = self._train_epoch(epoch)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/training/trainer.py", line 663, in _train_epoch
    for batch_group in batch_group_generator_tqdm:
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/tqdm/std.py", line 1133, in __iter__
    for obj in iterable:
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/common/util.py", line 139, in lazy_groups_of
    s = list(islice(iterator, group_size))
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/data/data_loaders/multiprocess_data_loader.py", line 333, in __iter__
    yield next(batch_generator)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/data/data_loaders/multiprocess_data_loader.py", line 395, in _iter_batches
    workers = self._start_batch_workers(queue, ctx)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/site-packages/allennlp/data/data_loaders/multiprocess_data_loader.py", line 434, in _start_batch_workers
    worker.start()
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/scratch/anaconda3/envs/devel/lib/python3.8/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'construct_arg.<locals>.constructor'
```
I just wanted to specify that I'm *not* creating a `TextField` specifying a token indexers and I followed your tutorial to currently implemented distributed and multiprocess applications (using `shard_iterable`).

**Comment by epwalsh** [comment]
@aleSuglia hm I think that error is because the vision reader stores these `Lazy`/`FromParams` objects:

https://github.com/allenai/allennlp-models/blob/3624fdc4b249d3b515269ee99e1b1e15bf876829/allennlp_models/vision/dataset_readers/vision_reader.py#L90-L91

And these objects internally have some locally defined functions, which can't be serialized by pickle. I'm looking into this more to see if there's a work-around.

**Comment by aleSuglia** [comment]
@epwalsh Yeah that makes sense. Because I tried in a simple example to instantiate directly the most specific classes (e.g. `ImageLoader`, `ResnetBackbone`, and `FasterRcnnRegionDetector`) and it works just fine. 
Also, any ideas about the reasons why the training hangs? It looks like I can get it to work when I set `batches_per_epoch` on the training. Any ideas as to how I can try to simplify the code so that I can share more useful information for you to debug? I guess my training data statistics didn't help?!

**Comment by epwalsh** [comment]
https://github.com/allenai/allennlp/pull/5049 should fix the pickling error.

> Any ideas as to how I can try to simplify the code so that I can share more useful information for you to debug?

Debugging multiprocessing code is hard. I would just throw logging / print statements in until I find the exact line that hangs. If you can find that line, I should be able to tell what the issue is.

**Comment by aleSuglia** [comment]
Mmmm even with the latest code release I get the following error:
```
AttributeError: Can't pickle local object 'ROUGE.__init__.<locals>.<lambda>'
```

**Comment by epwalsh** [comment]
https://github.com/allenai/allennlp/pull/5051 should ensure ROUGE can be pickled.

**Comment by dirkgr** [comment]
@aleSuglia, you have found quite a few issues here, and we fixed a fair number of them. Do you have any issues left now? I know about the `MDB_BAD_RSLOT` problem, but that already has its own issue. Can we close this one and track the rest separately?
